articles/WideStrings.md

```diff
@@ -0,0 +1,264 @@
+
+from std_msgs.msg import WString
+
+def main(args=None):
+    if args is None:
+        args = sys.argv
+
+    rclpy.init(args)
+
+    node = rclpy.create_node('talker')
+
+    chatter_pub = node.create_publisher(WString, 'chatter', qos_profile_default)
+
+    msg = String()
+
+    i = 1
+    while True:
+        msg.data = 'Hello World: {0}'.format(i)
+        i += 1
+        print('Publishing: "{0}"'.format(msg.data))
+        chatter_pub.publish(msg)
+        # TODO(wjwwood): need to spin_some or spin_once with timeout
+        sleep(1)
+
+
+if __name__ == '__main__':
+    main()
+```
+
+Notice that this code looks almost identical to the same demo for String.  That's because
+in python, the characters in the string and the byte representation are cleanly
+separated.  Thus, the API looks exactly the same as it would for the String type.
+
+#### C++
+
+C++ doesn't have good built-in support for wide strings.  Using wchar_t is generally
+not a good option, as wchar_t has different sizes on different platforms (2 bytes on
+Windows, 4 bytes on Linux).  Thus, ROS 2.0 will define a new type to deal with wide strings.
+The APIs that deal with wide strings will expect this new type to be passed in.
+
+**Example**
+
+```
+#include <iostream>
+#include <memory>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "std_msgs/msg/w_string.hpp"
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+
+  auto node = rclcpp::node::Node::make_shared("talker");
+
+  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
+  custom_qos_profile.depth = 7;
+
+  auto chatter_pub = node->create_publisher<std_msgs::msg::WString>("chatter", custom_qos_profile);
+
+  rclcpp::WallRate loop_rate(2);
+
+  auto msg = std::make_shared<std_msgs::msg::WString>();
+  auto i = 1;
+
+  while (rclcpp::ok()) {
+    utf8_string hello("Hello World: " + std::to_string(i++));
+    msg->data = hello;
+    std::cout << "Publishing: '" << msg->data.cpp_str() << "'" << std::endl;
+    chatter_pub->publish(msg);
+    rclcpp::spin_some(node);
+    loop_rate.sleep();
+  }
+
+  return 0;
+}
+```
+
+Here, ROS 2 has implemented a new type called "utf8_string" (this is still to be implemented,
+and the name may change).  The publisher takes in types of utf8_string, and generates the
+correct data on the wire.
+
+### Bounded wide strings
+
+A note on bounded wide strings.  ROS 2.0 allows for "bounded" strings and wide strings.  These
+are strings that can be no longer than their bounded number.  In the case of strings, the
+width of a character == the width of a byte in most cases, so the number of bytes in a bounded
+string is generally the bound (plus 1 for the NULL-termination).  Wide strings are a little bit different.
+When dealing with bounded wide strings, the bound is on the number of characters, not bytes.  Thus
+the maximum number of bytes the bounded wide string can be may be, and most likely will be, larger
+than the number of characters.
```

Comment:
It would be interesting to explore how preallocation would work in this case. The main 2 use cases for bounded strings are memory constrained system (microcontrollers) and realtime. One conclusion could be that for both these use cases we don't recommend using this type. Otherwise we would need to explore memory allocation strategies (e.g. allocate 4byte per character and we're happy if it takes less)